prov/verbs/src/verbs_msg.c

```diff
@@ -30,52 +30,104 @@
  * SOFTWARE.
  */
 
-#include "config.h"
+#include <rdma/rdma_verbs.h>
 
 #include "fi_verbs.h"
 
+static int
+fi_ibv_get_recv_comp(struct fid_ep *ep_fid)
+{
+	struct fi_ibv_msg_ep *ep = container_of(ep_fid, struct fi_ibv_msg_ep, ep_fid);
+	struct ibv_wc wc;
+	int ret;
+
+	while ((ret = rdma_get_recv_comp(ep->id, &wc)) == 0);
+	if (ret == 1 && wc.status == IBV_WC_SUCCESS)
+		return 0;
+	else
+		return -errno;
+}
+
+static int
+fi_ibv_get_send_comp(struct fid_ep *ep_fid)
+{
+	struct fi_ibv_msg_ep *ep = container_of(ep_fid, struct fi_ibv_msg_ep, ep_fid);
+	struct ibv_wc wc;
+	int ret;
+
+	while ((ret = rdma_get_send_comp(ep->id, &wc)) == 0);
+	if (ret == 1 && wc.status == IBV_WC_SUCCESS)
+		return 0;
+	else
+		return -errno;
+}
+
+static ssize_t
+fi_ibv_iov_total(size_t count, const struct iovec *iov)
+{
+	ssize_t total = 0;
+
+	for (int i = 0; i < count; ++i)
+		total += iov[i].iov_len;
+
+	assert(total >= 0);
+	return total;
+}
+
+static ssize_t
+fi_ibv_msg_total(const struct fi_msg *msg)
+{
+	return fi_ibv_iov_total(msg->iov_count, msg->msg_iov);
+}
 
 static ssize_t
 fi_ibv_msg_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flags)
 {
 	struct fi_ibv_msg_ep *_ep;
-	struct ibv_recv_wr wr, *bad;
-	struct ibv_sge *sge = NULL;
+	struct ibv_sge *sge;
 	ssize_t ret;
-	size_t i;
+
+	// fi_recvmsg() appears to not implement any flagged behavior
+	assert(flags == 0);
 
 	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
-	assert(_ep->rcq);
 
-	wr.wr_id = (uintptr_t) msg->context;
-	wr.next = NULL;
 	if (msg->iov_count) {
 		sge = alloca(sizeof(*sge) * msg->iov_count);
-		for (i = 0; i < msg->iov_count; i++) {
+		for (size_t i = 0; i < msg->iov_count; i++) {
 			sge[i].addr = (uintptr_t) msg->msg_iov[i].iov_base;
 			sge[i].length = (uint32_t) msg->msg_iov[i].iov_len;
 			sge[i].lkey = (uint32_t) (uintptr_t) (msg->desc[i]);
 		}
-
 	}
-	wr.sg_list = sge;
-	wr.num_sge = msg->iov_count;
 
-	ret = ibv_post_recv(_ep->id->qp, &wr, &bad);
+	ret = rdma_post_recvv(_ep->id, msg->context, sge, msg->iov_count);
 
 	switch (ret) {
-	case ENOMEM:
-		return -FI_EAGAIN;
 	case -1:
-		/* Deal with non-compliant libibverbs drivers which set errno
-		 * instead of directly returning the error value */
 		return (errno == ENOMEM) ? -FI_EAGAIN : -errno;
+	case 0:
 	default:
-		return -ret;
+		return 0;
 	}
 }
 
 static ssize_t
+fi_ibv_sstream_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flags)
+{
+	ssize_t ret;
+
+	ret = fi_ibv_msg_ep_recvmsg(ep, msg, flags);
+	if (ret)
+		return ret;
+
+	if ((ret = fi_ibv_get_recv_comp(ep)) == 0)
+		return fi_ibv_msg_total(msg);
```

Comment:
See comment below regarding memory registration and completion issues.

Additionally, this won't support nonblocking socket semantics.  The sender cannot transfer data until the receiver has actually posted the receive buffer.  Otherwise flow control will back the sender off until a buffer has been posted. The receive buffer must be made available to the sender prior to the application calling receive.  Recv() is used to retrieve locally buffered data, not provide network buffers.